jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/YenShortestPathIterator.java

```diff
@@ -0,0 +1,506 @@
+
+        double candidateWeight = rootPathWeight + spurPath.getWeight();
+        return new GraphWalk<>(graph, source, sink,
+                candidatePathVertices, candidatePathEdges, candidateWeight);
+    }
+
+    /**
+     * Checks if the lists have the same content until the {@code index} (inclusive).
+     *
+     * @param first  first list
+     * @param second second list
+     * @param index  position in the lists
+     * @return true iff the contents of the list are equal until the index
+     */
+    private boolean equalLists(List<V> first, List<V> second, int index) {
+        for (int i = 0; i <= index; i++) {
+            if (!first.get(i).equals(second.get(i))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Helper class which represents the shortest paths tree based on which
+     * the spur paths for the candidates are computed.
+     */
+    class YenShortestPathsTree extends TreeSingleSourcePathsImpl<V, E> {
+        /**
+         * Vertices which are masked in the {@link #g}.
+         */
+        Set<V> maskedVertices;
+        /**
+         * Edges which are masked in the {@link #g}.
+         */
+        Set<E> maskedEdges;
+
+        /**
+         * Constructs an instance of the shortest paths tree for the given
+         * {@code maskSubgraph}, {@code maskedVertices}, {@code maskedEdges}, {@code reversedTree}, {@code treeSource}.
+         *
+         * @param maskSubgraph   graph which has removed vertices and edges
+         * @param maskedVertices vertices removed form the graph
+         * @param maskedEdges    edges removed from the graph
+         * @param reversedTree   shortest path tree in the edge reverted
+         *                       {@code maskSubgraph} starting at {@code treeSource}.
+         * @param treeSource     source vertex of the  {@code reversedTree}
+         */
+        YenShortestPathsTree(Graph<V, E> maskSubgraph, Set<V> maskedVertices, Set<E> maskedEdges,
+                             Map<V, Pair<Double, E>> reversedTree, V treeSource) {
+            super(maskSubgraph, treeSource, reversedTree);
+            this.maskedVertices = maskedVertices;
+            this.maskedEdges = maskedEdges;
+        }
+
+        /**
+         * Restores vertex {@code v} in the {@link #g}.
+         *
+         * @param v vertex to be recovered
+         */
+        void recoverVertex(V v) {
+            maskedVertices.remove(v);
+        }
+
+        /**
+         * Restores edge {@code e} in the {@link #g}.
+         *
+         * @param e edge to be recovered
+         */
+        void recoverEdge(E e) {
+            maskedEdges.remove(e);
+        }
+
+        /**
+         * Updates the cost of provided vertex {@code v} in the shortest paths tree
+         * based on the current costs of its successors in the {@link #g}.
+         *
+         * @param v vertex which should be updated
+         */
+        void correctCostForward(V v) {
```

Comment:
Regarding the terminology: we have to call the edge weights either as edge distances or as edge costs. I would prefer using distances since it would be consistent with other shortest paths algorithm implementations